## Project 3 - Part 1

Author: Qichun Yu

### Question 1

a) Create a new worksheet and name it Project3YourName.<br>
b) Read in The Psy file, and display the tail

In [1]:
import pandas as pd

df = pd.read_csv("Youtube01-Psy.csv")

In [2]:
df.tail()

COMMENT_ID  \
345      z13th1q4yzihf1bll23qxzpjeujterydj   
346  z13fcn1wfpb5e51xe04chdxakpzgchyaxzo0k   
347    z130zd5b3titudkoe04ccbeohojxuzppvbg   
348    z12he50arvrkivl5u04cctawgxzkjfsjcc4   
349  z13vhvu54u3ewpp5h04ccb4zuoardrmjlyk0k   

                                   AUTHOR                 DATE  \
345                       Carmen Racasanu  2014-11-14T13:27:52   
346                       diego mogrovejo  2014-11-14T13:28:08   
347  BlueYetiPlayz -Call Of Duty and More  2015-05-23T13:04:32   
348                          Photo Editor  2015-06-05T14:14:48   
349                            Ray Benich  2015-06-05T18:05:16   

                                               CONTENT  CLASS  
345  How can this have 2 billion views when there's...      0  
346         I don't now why I'm watching this in 2014      0  
347  subscribe to me for call of duty vids and give...      1  
348  hi guys please my android photo editor downloa...      1  
349  The first billion viewed this because they tho...      0

c) Give the number of rows in the files; and using query give the number of spam and not spam comments.

In [3]:
df.shape

(350, 5)

In [4]:
df['CLASS'].value_counts()

1    175
0    175
Name: CLASS, dtype: int64

In [5]:
print(f"There are {df.shape[0]} rows in the files. There are {df['CLASS'].value_counts()[1]} spam comments and {df['CLASS'].value_counts()[0]} not spam comments")

There are 350 rows in the files. There are 175 spam comments and 175 not spam comments


### Question 2

a) Import the Bag of Words function and create an instance.

In [6]:
from sklearn.feature_extraction.text import CountVectorizer
vectorizer = CountVectorizer()

b) Fit and transform the comments in one step.

In [7]:
dvec = vectorizer.fit_transform(df['CONTENT'])

c) Output the fitted and transformed comments. You should have a sparse matrix.

In [8]:
dvec

<350x1418 sparse matrix of type '<class 'numpy.int64'>'
	with 4354 stored elements in Compressed Sparse Row format>

d) What are the dimensions of the matrix?

In [9]:
dvec.shape

(350, 1418)

The dimensions of the matrix is 350 x 1418. 

e) Print out the 349th comment 

In [10]:
print(df['CONTENT'][349])

The first billion viewed this because they thought it was really cool, the  other billion and a half came to see how stupid the first billion were...


f) Using vectorizer.build_analyzer() display the breakdown of the 349th comment in to a “bag of words”

In [11]:
analyze = vectorizer.build_analyzer()
analyze(df['CONTENT'][349])

['the',
 'first',
 'billion',
 'viewed',
 'this',
 'because',
 'they',
 'thought',
 'it',
 'was',
 'really',
 'cool',
 'the',
 'other',
 'billion',
 'and',
 'half',
 'came',
 'to',
 'see',
 'how',
 'stupid',
 'the',
 'first',
 'billion',
 'were']

g) Give the output of vectorizer.get_features_names(). What does this represent?

In [12]:
vectorizer.get_feature_names()

['00',
 '000',
 '02',
 '034',
 '05',
 '08',
 '10',
 '100',
 '100000415527985',
 '10200253113705769',
 '1030',
 '1073741828',
 '11',
 '1111',
 '112720997191206369631',
 '12',
 '123',
 '124',
 '124923004',
 '126',
 '127',
 '13017194',
 '131338190916',
 '1340488',
 '1340489',
 '1340490',
 '1340491',
 '1340492',
 '1340493',
 '1340494',
 '1340499',
 '1340500',
 '1340502',
 '1340503',
 '1340504',
 '1340517',
 '1340518',
 '1340519',
 '1340520',
 '1340521',
 '1340522',
 '1340523',
 '1340524',
 '134470083389909',
 '1415297812',
 '1495323920744243',
 '1496241863981208',
 '1496273723978022',
 '1498561870415874',
 '161620527267482',
 '171183229277',
 '19',
 '19924',
 '1firo',
 '1m',
 '20',
 '2009',
 '2012',
 '2012bitches',
 '2013',
 '2014',
 '201470069872822',
 '2015',
 '2017',
 '210',
 '23',
 '24',
 '24398',
 '243a',
 '279',
 '29',
 '2b',
 '2billion',
 '2x10',
 '300',
 '3000',
 '313327',
 '315',
 '322',
 '33',
 '33gxrf',
 '39',
 '390875584405933',
 '391725794320912',
 '40beuutvu2zkxk4utgpz8k',
 '

Each one is a tokenized word. The function returns a list of feature names selected from the raw documents. We can identify words that were found after vectorizing.

### Question 3

a) Shuffle the dataset (frac=1).

In [13]:
df = df.sample(frac=1)

In [14]:
df.head()

COMMENT_ID            AUTHOR  \
181  z13neh044nq2wn0o404citkh2taufzqxuz40k  Mason Sieverding   
76     z13gsp34uumcdje4q04civk54zr4wr0gxjc          Lucky D.   
166      z13mebgjooulgzdkt23kzvlqjyrzx32n0        amine moha   
147    z12zstfhixudundzf04cjbsp0rjrzh1rw3k        MiningBip3   
125      z124i3ygclj0xn4ly22jzb5a2ubvtbev1    Abdinasir Omar   

                    DATE                                            CONTENT  \
181  2014-11-07T01:50:10                        Please check out my vidios   
76   2014-11-02T14:53:06         http://hackfbaccountlive.com/?ref=4604617   
166  2014-11-06T18:11:37  please like : http://www.bubblews.com/news/927...   
147  2014-11-06T02:22:00          Check out pivot animations in my channel   
125  2014-11-05T15:28:28                                Great music anyway   

     CLASS  
181      1  
76       1  
166      1  
147      1  
125      0

b) Create your training and testing sets. Use the first 300 entries for training and the remaining for testing. Name the files d_train, and d_test.

In [15]:
d_train = df[:300]
d_test = df[300:]

c) Create your training and testing attributes BOW using using vectorizer.fit_transform. Name these d_train_att, and d_test_att.

In [16]:
d_train_att = vectorizer.fit_transform(d_train['CONTENT'])
d_test_att = vectorizer.transform(d_test['CONTENT'])

d) Create your training and testing labels. Name these d_train_label, and d_test_label.

In [17]:
d_train_label = d_train['CLASS']
d_test_label = d_test['CLASS']

e) Output both d_train_att, and d_test_att. What are the dimensions of the matrices?

In [18]:
d_train_att

<300x1288 sparse matrix of type '<class 'numpy.int64'>'
	with 3753 stored elements in Compressed Sparse Row format>

The dimensions of the d_train_att matrice is 300x1288.

In [19]:
d_test_att

<50x1288 sparse matrix of type '<class 'numpy.int64'>'
	with 470 stored elements in Compressed Sparse Row format>

The dimensions of the d_test_att matrice is 50x1288. 

### Question 4

a) Crate a random forest classifier with 80 trees. Output.

In [20]:
from sklearn.ensemble import RandomForestClassifier

clf = RandomForestClassifier(n_estimators=80)

In [21]:
clf.fit(d_train_att, d_train_label)

RandomForestClassifier(bootstrap=True, ccp_alpha=0.0, class_weight=None,
                       criterion='gini', max_depth=None, max_features='auto',
                       max_leaf_nodes=None, max_samples=None,
                       min_impurity_decrease=0.0, min_impurity_split=None,
                       min_samples_leaf=1, min_samples_split=2,
                       min_weight_fraction_leaf=0.0, n_estimators=80,
                       n_jobs=None, oob_score=False, random_state=None,
                       verbose=0, warm_start=False)

b) Using .score, score your tree.

In [22]:
clf.score(d_test_att, d_test_label)

0.98

c) Create a confusion matrix for the test labels and prediction labels, and out put his array.

In [23]:
from sklearn.metrics import confusion_matrix

predicted_labels = clf.predict(d_test_att)
cm = confusion_matrix(d_test_label, predicted_labels)

In [24]:
cm

array([[28,  0],
       [ 1, 21]], dtype=int64)

d) Cross validate and output your accuracy, i.e., scores.mean +/- 2sd. Make sure to use your training attributes and training labels here.Set cv=5

In [25]:
from sklearn.model_selection import cross_val_score

scores_RF = cross_val_score(clf, d_train_att, d_train_label, cv=5)

In [26]:
print("Accuracy of the Random Forest: %0.2f (+/- %0.2f)"%(scores_RF.mean(), scores_RF.std() * 2))

Accuracy of the Random Forest: 0.97 (+/- 0.04)


e) In your words explain the cross validation step, and what it is used for. You may use any references you wish. Please include any references you use.

The first step of cross-validation is to split the data into subsets(or folds). Then, it will treat each subset as a training set or a validating set. For example, when we set the cv=5, the dataset will be divided into 5 folds. Then, it will experiment five times where the first time treats the first fold as a validating set and the remaining four folds as training. Each experiment will split the data as 80% training and 20% validation. We repeat the experiment five times until each fold has been a validation set one time. 

||fold 1| fold 2 | fold 3 |fold 4|fold 5|
|:-| :-: | :-: | :-: | :-: | :-: |
|Experiment 1 | Validation | Training | Training | Training | Training |
|Experiment 2 | Training | Validation| Training | Training | Training |
|Experiment 3 | Training | Training | Validation| Training | Training |
|Experiment 4 | Training | Training | Training | Validation | Training |
|Experiment 5 | Training | Training | Training | Training | Validation|

The cross-validation is used for validating the machine learning model to ensure its accuracy and measure its performance. The reason to validate the machine learning model is to make sure the model is able to perform well on unseen data. The cross-validation usually provides a more accurate measurement of the performance of the model because it allows the model to train and test on different splits. 

__References:__<br>
3.1. Cross-validation: evaluating estimator performance. (n.d.). Scikit-learn. Retrieved October 25, 2022, from https://scikit-learn.org/stable/modules/cross_validation.html

### Question 5

a) Load and concatenate all 5 comments files.

In [27]:
data = pd.concat(map(pd.read_csv, ['Youtube01-Psy.csv', 
                                   'Youtube02-KatyPerry.csv', 
                                   'Youtube03-LMFAO.csv',
                                   'Youtube04-Eminem.csv',
                                   'Youtube05-Shakira.csv']))

In [28]:
data.head()

COMMENT_ID            AUTHOR  \
0  LZQPQhLyRh80UYxNuaDWhIGQYNQ96IuCg-AYWqNPjpU         Julius NM   
1  LZQPQhLyRh_C2cTtd9MvFRJedxydaVW-2sNg5Diuo4A       adam riyati   
2  LZQPQhLyRh9MSZYnf8djyk0gEF9BHDPYrrK-qCczIY8  Evgeny Murashkin   
3          z13jhp0bxqncu512g22wvzkasxmvvzjaz04   ElNino Melendez   
4          z13fwbwp1oujthgqj04chlngpvzmtt3r3dw            GsMega   

                  DATE                                            CONTENT  \
0  2013-11-07T06:20:48  Huh, anyway check out this you[tube] channel: ...   
1  2013-11-07T12:37:15  Hey guys check out my new channel and our firs...   
2  2013-11-08T17:34:21             just for test I have to say murdev.com   
3  2013-11-09T08:28:43   me shaking my sexy ass on my channel enjoy ^_^    
4  2013-11-10T16:05:38            watch?v=vtaRGgvGtWQ   Check this out .   

   CLASS  
0      1  
1      1  
2      1  
3      1  
4      1

b) Display the length of the file as well as the breakdown into spam and non-spam.

In [29]:
# length of the file
len(data)

1956

In [30]:
# length of the spam
len(data[data['CLASS']==1])

1005

In [31]:
# length of the non-spam
len(data[data['CLASS']==0])

951

c) Shuffle the new data, and create content and label sets, d_content, and d_label respectively.

In [32]:
data = data.sample(frac=1)

data.head()

COMMENT_ID               AUTHOR  \
261  z13shj4wpmflidcxc04ce5f4vlqdyzjowso0k  Tornike Noniashvili   
40     z13lhdo5vnyndj3p304cjjs4zze3sldrqbg           Mai Nguyễn   
347      z132jbmxfqm4fjysg23nwjfb2mv2vxnua  Decio Alves Martins   
313      z12mdpxzzvmxwfevl23juhzibvqyvp52s         Martwy Karas   
203    z12xizopmzbch3dje04cibrzbqi2x54ptgk     Shaina Rodriguez   

                           DATE  \
261         2014-11-08T04:08:09   
40          2014-08-28T03:53:35   
347         2015-06-05T19:29:20   
313         2014-11-07T19:33:46   
203  2015-05-16T00:20:41.387000   

                                               CONTENT  CLASS  
261                               subscribe my chanel      1  
40   help me click on the subscribe Mai Nguyen, tha...      1  
347  KATY PERRY, I AM THE "DÉCIO CABELO", "DECIO HA...      1  
313  Katy Perry's songs are the best of the songs o...      0  
203                              lol so funny love it      0

In [33]:
d_content = data['CONTENT']
d_label = data['CLASS']

### Question 6

a) Read the scikit learn documentation on Pipelines. What is an advantage to using a pipeline? Use any reference you wish to answer the question, just be sure to include the reference in your response.

Pipeline allows to perform cross-validation while setting different parameters for the model. The advantages to use a pipeline are:
<ul>
    <li>Machine learning workflow will be easy to read and understand</li>
    <li>Enforce the implementation and order of steps in your model</li>
    <li>Easy to reproduce</li>
</ul>

__References:__<br>
sklearn.pipeline.Pipeline. (n.d.). Scikit-learn. Retrieved October 25, 2022, from https://scikit-learn.org/stable/modules/generated/sklearn.pipeline.Pipeline.html<br>

Vickery, R. (2021, December 7). A Simple Guide to Scikit-learn Pipelines - vickdata. Medium. https://medium.com/vickdata/a-simple-guide-to-scikit-learn-pipelines-4ac0d974bdcf

b) Import both Pipeline and make_pipeline from sklearn.

In [34]:
from sklearn.pipeline import Pipeline, make_pipeline

c) Create a pipeline. You may use the following code.

In [35]:
pipeline = Pipeline([
    ('bag-of-word', CountVectorizer()),
    ('random forest', RandomForestClassifier())
])

d) Output the pipeline.

In [36]:
pipeline

Pipeline(memory=None,
         steps=[('bag-of-word',
                 CountVectorizer(analyzer='word', binary=False,
                                 decode_error='strict',
                                 dtype=<class 'numpy.int64'>, encoding='utf-8',
                                 input='content', lowercase=True, max_df=1.0,
                                 max_features=None, min_df=1,
                                 ngram_range=(1, 1), preprocessor=None,
                                 stop_words=None, strip_accents=None,
                                 token_pattern='(?u)\\b\\w\\w+\\b',
                                 tokenizer=None, vocabul...
                 RandomForestClassifier(bootstrap=True, ccp_alpha=0.0,
                                        class_weight=None, criterion='gini',
                                        max_depth=None, max_features='auto',
                                        max_leaf_nodes=None, max_samples=None,
                                

In [37]:
# or: pipeline = make_pipeline(CountVectorizer(), RandomForestClassifier())
make_pipeline(CountVectorizer(), RandomForestClassifier())

Pipeline(memory=None,
         steps=[('countvectorizer',
                 CountVectorizer(analyzer='word', binary=False,
                                 decode_error='strict',
                                 dtype=<class 'numpy.int64'>, encoding='utf-8',
                                 input='content', lowercase=True, max_df=1.0,
                                 max_features=None, min_df=1,
                                 ngram_range=(1, 1), preprocessor=None,
                                 stop_words=None, strip_accents=None,
                                 token_pattern='(?u)\\b\\w\\w+\\b',
                                 tokenizer=None, voc...
                 RandomForestClassifier(bootstrap=True, ccp_alpha=0.0,
                                        class_weight=None, criterion='gini',
                                        max_depth=None, max_features='auto',
                                        max_leaf_nodes=None, max_samples=None,
                                

e) Fit your pipeline with the first 1500 entries of the content and the labels. Output.

In [38]:
pipeline.fit(d_content[:1500], d_label[:1500])

Pipeline(memory=None,
         steps=[('bag-of-word',
                 CountVectorizer(analyzer='word', binary=False,
                                 decode_error='strict',
                                 dtype=<class 'numpy.int64'>, encoding='utf-8',
                                 input='content', lowercase=True, max_df=1.0,
                                 max_features=None, min_df=1,
                                 ngram_range=(1, 1), preprocessor=None,
                                 stop_words=None, strip_accents=None,
                                 token_pattern='(?u)\\b\\w\\w+\\b',
                                 tokenizer=None, vocabul...
                 RandomForestClassifier(bootstrap=True, ccp_alpha=0.0,
                                        class_weight=None, criterion='gini',
                                        max_depth=None, max_features='auto',
                                        max_leaf_nodes=None, max_samples=None,
                                

f) Use .score to score your pipeline.

In [39]:
pipeline.score(d_content[1500:], d_label[1500:])

0.9649122807017544

g) Use your pipeline to predict whether the following two comments are spam or Not spam. <br>
I. “what a neat video!”<br>
II. “plz subscribe to my channel”

In [40]:
pipeline.predict(["what a neat video!"])

array([0], dtype=int64)

It returns 0, so it is not a spam.

In [41]:
pipeline.predict(["plz subscribe to my channel"])

array([1], dtype=int64)

It returns 1, so it is a spam. 

### Question 7

a) Cross validate your pipeline using d_content and d_label. Set cv=5.

In [42]:
scores_pipeline = cross_val_score(pipeline, d_content, d_label, cv=5)

b) Print out the accuracy(scores.mean) +/- 2sd.

In [43]:
print("Accuracy of the pipeline: %0.2f (+/- %0.2f)"%(scores_pipeline.mean(), scores_pipeline.std() * 2))

Accuracy of the pipeline: 0.96 (+/- 0.02)


### Question 8

a) Create a second pipeline named pipeline2 which incorporates the TfidfTransformer

In [44]:
# add tfidf
from sklearn.feature_extraction.text import TfidfTransformer
pipeline2 = make_pipeline(CountVectorizer(),
                          TfidfTransformer(norm=None),
                          RandomForestClassifier())

b) Cross validate pipeline2 using d_content and d_labels.Output the accuracy(scores.mean) +/- 2sd. Set cv=5.

In [45]:
scores_pipeline2 = cross_val_score(pipeline2, d_content, d_label, cv=5)

In [46]:
print("Accuracy of the pipeline2: %0.2f (+/- %0.2f)"%(scores_pipeline2.mean(), scores_pipeline2.std() * 2))

Accuracy of the pipeline2: 0.96 (+/- 0.01)


c) Output the steps of pipeline 2.

In [47]:
pipeline2.steps

[('countvectorizer',
  CountVectorizer(analyzer='word', binary=False, decode_error='strict',
                  dtype=<class 'numpy.int64'>, encoding='utf-8', input='content',
                  lowercase=True, max_df=1.0, max_features=None, min_df=1,
                  ngram_range=(1, 1), preprocessor=None, stop_words=None,
                  strip_accents=None, token_pattern='(?u)\\b\\w\\w+\\b',
                  tokenizer=None, vocabulary=None)),
 ('tfidftransformer',
  TfidfTransformer(norm=None, smooth_idf=True, sublinear_tf=False, use_idf=True)),
 ('randomforestclassifier',
  RandomForestClassifier(bootstrap=True, ccp_alpha=0.0, class_weight=None,
                         criterion='gini', max_depth=None, max_features='auto',
                         max_leaf_nodes=None, max_samples=None,
                         min_impurity_decrease=0.0, min_impurity_split=None,
                         min_samples_leaf=1, min_samples_split=2,
                         min_weight_fraction_leaf=0.0, 

### Question 9

a) Use the following code snippet to set up a grid search.

In [48]:
# parameter search
parameters = {
    'countvectorizer__max_features': (None, 1000, 2000),
    'countvectorizer__ngram_range': ((1,1), (1, 2)), # unigrams or bigrams
    'countvectorizer__stop_words': ('english', None),
    'tfidftransformer__use_idf': (True, False), # effectively turn on/off tfidf
    'randomforestclassifier__n_estimators': (20, 50, 100)
}
from sklearn.model_selection import GridSearchCV
grid_search = GridSearchCV(pipeline2, parameters, n_jobs=-1, verbose=1)

b) Describe the purpose of the code in part (a) within the context of the grid search. What is this setting up?

The grid search allows the user to specify different hyperparameters and values and calculates the performance for each combination and find the optimal performance of those hyperparameters settings. The first thing within the context is the estimator, which in our case is pipeline2, it is the model that we want to apply. The second thing within the context of the grid search is the param_grid, which is the parameter that we defined to store all of the values of the hyperparameters with parameter names as the key. The n_jobs is the number of jobs that we want to run in parallel, -1 means using all processors. The last one verbose controls the verbosity, a higher value indicates more messages will be displayed. 

c) Using .fit nd d_content, and d_labels, perform the grid search.

In [49]:
grid_search.fit(d_content, d_label)

Fitting 5 folds for each of 72 candidates, totalling 360 fits


[Parallel(n_jobs=-1)]: Using backend LokyBackend with 12 concurrent workers.
[Parallel(n_jobs=-1)]: Done  26 tasks      | elapsed:    3.1s
[Parallel(n_jobs=-1)]: Done 176 tasks      | elapsed:    8.5s
[Parallel(n_jobs=-1)]: Done 360 out of 360 | elapsed:   13.7s finished


GridSearchCV(cv=None, error_score=nan,
             estimator=Pipeline(memory=None,
                                steps=[('countvectorizer',
                                        CountVectorizer(analyzer='word',
                                                        binary=False,
                                                        decode_error='strict',
                                                        dtype=<class 'numpy.int64'>,
                                                        encoding='utf-8',
                                                        input='content',
                                                        lowercase=True,
                                                        max_df=1.0,
                                                        max_features=None,
                                                        min_df=1,
                                                        ngram_range=(1, 1),
                                              

d) Print out the following.

i. Best score<br>
ii. Best countvectorizer_max_features value<br>
iii. Best countvectorizer_ngram_range value<br>
iv. Stop words setting<br>
v. Best n_estimators setting value<br>
vi. TF-IDF setting<br>

In [50]:
grid_search.best_estimator_

Pipeline(memory=None,
         steps=[('countvectorizer',
                 CountVectorizer(analyzer='word', binary=False,
                                 decode_error='strict',
                                 dtype=<class 'numpy.int64'>, encoding='utf-8',
                                 input='content', lowercase=True, max_df=1.0,
                                 max_features=1000, min_df=1,
                                 ngram_range=(1, 2), preprocessor=None,
                                 stop_words=None, strip_accents=None,
                                 token_pattern='(?u)\\b\\w\\w+\\b',
                                 tokenizer=None, voc...
                 RandomForestClassifier(bootstrap=True, ccp_alpha=0.0,
                                        class_weight=None, criterion='gini',
                                        max_depth=None, max_features='auto',
                                        max_leaf_nodes=None, max_samples=None,
                                

In [51]:
grid_search.best_params_

{'countvectorizer__max_features': 1000,
 'countvectorizer__ngram_range': (1, 2),
 'countvectorizer__stop_words': None,
 'randomforestclassifier__n_estimators': 100,
 'tfidftransformer__use_idf': False}

In [52]:
print("Best score: {0:.3f}".format(grid_search.best_score_))
print("Best parameters set:")
for parameter, value in grid_search.best_params_.items():
    print(f"\t{parameter}: {value}")

Best score: 0.962
Best parameters set:
	countvectorizer__max_features: 1000
	countvectorizer__ngram_range: (1, 2)
	countvectorizer__stop_words: None
	randomforestclassifier__n_estimators: 100
	tfidftransformer__use_idf: False
